<h1>Game Analysis - Integrated Project 1</h1>

In [1]:
# import streamlit as st
import pandas as pd
import scipy.stats as stats
import numpy as np
# import plotly.express as px

ModuleNotFoundError: No module named 'scipy'

In [ ]:
df = pd.read_csv('games.csv') #import your dataset

print(df.describe(include='object'))#describe dataset
df.info()#info from dataset
display(df.head(10))
display(df.sample(10))
display(df.tail(10))#sample data

                               Name Platform   Genre User_Score Rating
count                         16713    16715   16713      10014   9949
unique                        11559       31      12         96      8
top     Need for Speed: Most Wanted      PS2  Action        tbd      E
freq                             12     2161    3369       2424   3990
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16713 non-null  object 
 1   Platform         16715 non-null  object 
 2   Year_of_Release  16446 non-null  float64
 3   Genre            16713 non-null  object 
 4   NA_sales         16715 non-null  float64
 5   EU_sales         16715 non-null  float64
 6   JP_sales         16715 non-null  float64
 7   Other_sales      16715 non-null  float64
 8   Critic_Score     8137 non-null   float64
 9   User_Score       10014 no

Name Platform  Year_of_Release         Genre  \
0                 Wii Sports      Wii           2006.0        Sports   
1          Super Mario Bros.      NES           1985.0      Platform   
2             Mario Kart Wii      Wii           2008.0        Racing   
3          Wii Sports Resort      Wii           2009.0        Sports   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
6      New Super Mario Bros.       DS           2006.0      Platform   
7                   Wii Play      Wii           2006.0          Misc   
8  New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                  Duck Hunt      NES           1984.0       Shooter   

   NA_sales  EU_sales  JP_sales  Other_sales  Critic_Score User_Score Rating  
0     41.36     28.96      3.77         8.45          76.0          8      E  
1     29.08      3.58      6.81         0.77           NaN        NaN    NaN  
2     15.68     12.76      3.79         3.29          82.0        8.3      E  
3     15.61     10.93      3.28         2.95          80.0          8      E  
4     11.27      8.89     10.22         1.00           NaN        NaN    NaN  
5     23.20      2.26      4.22         0.58           NaN        NaN    NaN  
6     11.28      9.14      6.50         2.88          89.0        8.5      E  
7     13.96      9.18      2.93         2.84          58.0        6.6      E  
8     14.44      6.94      4.70         2.24          87.0        8.4      E  
9     26.93      0.63      0.28         0.47           NaN        NaN    NaN

Name Platform  Year_of_Release  \
12180                               Mario Tennis      Wii           2010.0   
10661                         Ghost in the Shell       PS           1997.0   
585                               FIFA Soccer 10     X360           2009.0   
653             Halo: Combat Evolved Anniversary     X360           2011.0   
10173                           Guitar Rock Tour       DS           2007.0   
8452           Cubivore: Survival of the Fittest       GC           2002.0   
6374   Jikkyou Powerful Pro Yakyuu '98 Ketteiban       PS           1998.0   
2659                        Conker's Bad Fur Day      N64           2001.0   
924                High School Musical: Sing It!      PS2           2007.0   
1284                    Disney Sing It: Pop Hits      Wii           2009.0   

            Genre  NA_sales  EU_sales  JP_sales  Other_sales  Critic_Score  \
12180      Sports      0.00      0.06      0.00         0.01           NaN   
10661  Simulation      0.06      0.04      0.00         0.01           NaN   
585        Sports      0.59      1.79      0.01         0.23          90.0   
653       Shooter      1.44      0.70      0.04         0.22          82.0   
10173        Misc      0.10      0.00      0.00         0.01          60.0   
8452    Adventure      0.01      0.00      0.15         0.00          71.0   
6374       Sports      0.00      0.00      0.25         0.02           NaN   
2659     Platform      0.53      0.22      0.00         0.02           NaN   
924          Misc      0.47      0.08      0.00         1.33           NaN   
1284         Misc      1.07      0.28      0.00         0.13           NaN   

      User_Score Rating  
12180        tbd      E  
10661        NaN    NaN  
585          7.5      E  
653          8.1      M  
10173        tbd   E10+  
8452         8.5      E  
6374         NaN    NaN  
2659         NaN    NaN  
924          NaN    NaN  
1284         tbd      E

Name Platform  \
16705                                           15 Days       PC   
16706                     Men in Black II: Alien Escape       GC   
16707                                  Aiyoku no Eustia      PSV   
16708                Woody Woodpecker in Crazy Castle 5      GBA   
16709  SCORE International Baja 1000: The Official Game      PS2   
16710                     Samurai Warriors: Sanada Maru      PS3   
16711                                  LMA Manager 2007     X360   
16712                           Haitaka no Psychedelica      PSV   
16713                                  Spirits & Spells      GBA   
16714                               Winning Post 8 2016      PSV   

       Year_of_Release       Genre  NA_sales  EU_sales  JP_sales  Other_sales  \
16705           2009.0   Adventure      0.00      0.01      0.00          0.0   
16706           2003.0     Shooter      0.01      0.00      0.00          0.0   
16707           2014.0        Misc      0.00      0.00      0.01          0.0   
16708           2002.0    Platform      0.01      0.00      0.00          0.0   
16709           2008.0      Racing      0.00      0.00      0.00          0.0   
16710           2016.0      Action      0.00      0.00      0.01          0.0   
16711           2006.0      Sports      0.00      0.01      0.00          0.0   
16712           2016.0   Adventure      0.00      0.00      0.01          0.0   
16713           2003.0    Platform      0.01      0.00      0.00          0.0   
16714           2016.0  Simulation      0.00      0.00      0.01          0.0   

       Critic_Score User_Score Rating  
16705          63.0        5.8    NaN  
16706           NaN        tbd      T  
16707           NaN        NaN    NaN  
16708           NaN        NaN    NaN  
16709           NaN        NaN    NaN  
16710           NaN        NaN    NaN  
16711           NaN        NaN    NaN  
16712           NaN        NaN    NaN  
16713           NaN        NaN    NaN  
16714           NaN        NaN    NaN

Decribe here what you notice about the dataset: missing, duplicated, data types, etc.

This dataset contains information about video games. The dataset consists of 16,715 entries and 11 columns, including Name, Platform, Year_of_Release, Genre, NA_sales, EU_sales, JP_sales, Other_sales, Critic_Score, User_Score, and Rating. The data types are a mix of float64 (6 columns) and object (5 columns). There are missing values in several columns: Year_of_Release has 16,446 non-null entries (269 missing), Critic_Score has 8,137 non-null entries (8,578 missing), User_Score has 10,014 non-null entries (6,701 missing), and Rating has 9,949 non-null entries (6,766 missing). The dataset does not appear to have explicit duplicate rows, as each game title, platform, and year combination is unique; however, some names might be repeated with different platforms.

In [ ]:
df.columns = df.columns.str.lower() #make columns lower case
print(df.columns)

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')


In [ ]:
missing_percentage = df.isnull().mean() * 100 #missing values percentage
missing_values = df.isnull().sum()#missing values and how to handle them
print(missing_values)
print(missing_percentage)


name                  2
platform              0
year_of_release     269
genre                 2
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8578
user_score         6701
rating             6766
dtype: int64
name                0.011965
platform            0.000000
year_of_release     1.609333
genre               0.011965
na_sales            0.000000
eu_sales            0.000000
jp_sales            0.000000
other_sales         0.000000
critic_score       51.319174
user_score         40.089740
rating             40.478612
dtype: float64


It looks like the missing values in Year_of_Release, Critic_Score, and User_Score and Rating columns are significant. We will handle these missing values by replacing them with the mean value for the respective columns and 'tbd' for the scores. The rating column I will use the mode value to fill in the missing values because the mode is a more robust measure of central tendency than the mean.

In [ ]:
df['user_score'].replace('tbd', np.nan, inplace=True)
df['user_score'] = pd.to_numeric(df['user_score'], errors='coerce') #replace 'tbd' with NaN and convert to numeric
df['user_score'].fillna(df['user_score'].mean(), inplace=True)#handle missing values in user_score column by replacing 'tbd' with NaN and converting to numeric.
df['user_score'] = df['user_score'].astype(float) #convert user_score column to float type
print(df['user_score'].isnull().sum()) #check missing values in user_score

df['critic_score'].replace('tbd', np.nan, inplace=True)
df['critic_score'] = pd.to_numeric(df['critic_score'], errors='coerce') #replace 'tbd' with NaN and convert to numeric
df['critic_score'].fillna(df['critic_score'].mean(), inplace=True) #critic_score missing values with mean for critic_score column
df['critic_score'] = df['critic_score'].astype(float) #convert critic_score to float type
print(df['critic_score'].isnull().sum()) #check missing values in critic_score column
print(df.dtypes) #check data types after conversion

0
0
name                object
platform            object
year_of_release    float64
genre               object
na_sales           float64
eu_sales           float64
jp_sales           float64
other_sales        float64
critic_score       float64
user_score         float64
rating              object
dtype: object


I converted user_score and critic_score to float type, as the user_score was an object type, and 'tbd' was used to indicate that the score was not yet available.

In [ ]:
df['year_of_release'] = pd.to_datetime(df['year_of_release'], format='%Y', errors='coerce') #convert year_of_release to datetime and handle

df.dropna(subset=['year_of_release'], inplace=True) #drop missing values from year_of_release column

df['rating'].fillna(df['rating'].mode()[0], inplace=True) #fill missing values with mode for rating column

print(df['year_of_release'].isnull().sum())#check missing values
print(df.dtypes) #check data types after handling missing values and conversion

0
name                       object
platform                   object
year_of_release    datetime64[ns]
genre                      object
na_sales                  float64
eu_sales                  float64
jp_sales                  float64
other_sales               float64
critic_score              float64
user_score                float64
rating                     object
dtype: object


I converted the year_of_release column to datetime type and handled missing values by dropping rows with missing values. I also filled missing values in the rating column with the mode value.

In [ ]:
df_duplicates = df[df.duplicated()]#duplicated values and how to handle them
df_specific_columns = df[df.duplicated(subset=['name', 'platform', 'year_of_release'])]#specific

print(df_duplicates) #print duplicated values
print(df_specific_columns) #print specific columns

Empty DataFrame
Columns: [name, platform, year_of_release, genre, na_sales, eu_sales, jp_sales, other_sales, critic_score, user_score, rating]
Index: []
                name platform year_of_release   genre  na_sales  eu_sales  \
14244            NaN      GEN      1993-01-01     NaN       0.0      0.00   
16230  Madden NFL 13      PS3      2012-01-01  Sports       0.0      0.01   

       jp_sales  other_sales  critic_score  user_score rating  
14244      0.03          0.0     68.967679    7.125046      E  
16230      0.00          0.0     83.000000    5.500000      E  


In [ ]:
df_aggregate = df.groupby(['name', 'platform','year_of_release', 'genre', 'rating']).agg({'na_sales':'sum', 'eu_sales':'sum', 'jp_sales':'sum', 'other_sales':'sum', 'critic_score':'mean', 'user_score':'mean'}).reset_index()

print(df_aggregate) #print aggregated values

                               name platform year_of_release         genre  \
0                    Beyblade Burst      3DS      2016-01-01  Role-Playing   
1                 Fire Emblem Fates      3DS      2015-01-01  Role-Playing   
2              Frozen: Olaf's Quest      3DS      2013-01-01      Platform   
3              Frozen: Olaf's Quest       DS      2013-01-01      Platform   
4        Haikyu!! Cross Team Match!      3DS      2016-01-01     Adventure   
...                             ...      ...             ...           ...   
16438    thinkSMART: Chess for Kids       DS      2011-01-01          Misc   
16439                  uDraw Studio      Wii      2010-01-01          Misc   
16440  uDraw Studio: Instant Artist      Wii      2011-01-01          Misc   
16441  uDraw Studio: Instant Artist     X360      2011-01-01          Misc   
16442  ¡Shin Chan Flipa en colores!       DS      2007-01-01      Platform   

      rating  na_sales  eu_sales  jp_sales  other_sales  critic

I decided to handle the duplicates by aggregating the values in the columns where duplicates occur. This is because I believe that its the best way to ensure the most comprehensive view of the data.

In [ ]:
df_aggregate['total_sales'] = df_aggregate[['na_sales','eu_sales','jp_sales', 'other_sales']].sum(axis=1)#calculate the total sales in all regions and add it as a new column

display(df_aggregate.sample(10)) #print aggregated values with total sales column


name platform year_of_release  \
9149                            NCAA Football 2002      PS2      2001-01-01   
745    Atelier Shallie: Alchemists of the Dusk Sea      PS3      2014-01-01   
1905                                      Carnival     2600      1981-01-01   
14494      Theatrhythm Final Fantasy: Curtain Call      3DS      2014-01-01   
320                                      Alan Wake       PC      2012-01-01   
12264                   Sid Meier's Civilization V       PC      2010-01-01   
4094                      F.E.A.R. Perseus Mandate       PC      2007-01-01   
33          1000 Cooking Recipes from ELLE à table       DS      2010-01-01   
4572             Final Fantasy XIV: A Realm Reborn      PS3      2013-01-01   
9367                                  Namco Museum      PS2      2001-01-01   

              genre rating  na_sales  eu_sales  jp_sales  other_sales  \
9149         Sports      E      0.77      0.60      0.00         0.20   
745    Role-Playing   E10+      0.06      0.02      0.08         0.02   
1905        Shooter      E      0.32      0.02      0.00         0.00   
14494          Misc      T      0.13      0.08      0.15         0.02   
320          Action      T      0.00      0.06      0.00         0.01   
12264      Strategy   E10+      0.98      0.52      0.00         0.17   
4094        Shooter      M      0.01      0.00      0.00         0.00   
33             Misc      E      0.00      0.02      0.00         0.00   
4572   Role-Playing      E      0.37      0.30      0.28         0.13   
9367           Misc      E      1.73      0.07      0.00         0.02   

       critic_score  user_score  total_sales  
9149      90.000000    8.700000         1.57  
745       76.000000    7.700000         0.18  
1905      68.967679    7.125046         0.34  
14494     83.000000    8.200000         0.38  
320       83.000000    8.000000         0.07  
12264     90.000000    7.900000         1.67  
4094      61.000000    7.300000         0.01  
33        68.967679    7.125046         0.02  
4572      68.967679    7.125046         1.08  
9367      72.000000    6.000000         1.82

<h2> Analyze Data</h2>

<i>Look at how many games were released in different years. Is the data for every period significant?</i>

I will first look at the number of games released each year by parsing the year_of_release column. Then I will count the number of games released each year. Then use statistical tests to determine if the data for every period is significant.

Null: Hypothosis: The number of games released each year is not significantly different from each other.

Hypothesis: The number of games released each year is significantly different from each other.



In [ ]:
game_release_counts = df_aggregate['year_of_release'].dt.year.value_counts().sort_index() #count the number of games released each year and sort by year

print(game_release_counts) #print the number of games released each year

year_of_release
1980       9
1981      46
1982      36
1983      17
1984      14
1985      14
1986      21
1987      16
1988      15
1989      17
1990      16
1991      41
1992      43
1993      60
1994     121
1995     219
1996     263
1997     289
1998     379
1999     338
2000     350
2001     482
2002     829
2003     775
2004     762
2005     939
2006    1006
2007    1197
2008    1427
2009    1426
2010    1255
2011    1136
2012     652
2013     544
2014     581
2015     606
2016     502
Name: count, dtype: int64


Look at how sales varied from platform to platform. Choose the platforms with the greatest total sales and build a distribution based on data for each year. Find platforms that used to be popular but now have zero sales. How long does it generally take for new platforms to appear and old ones to fade?

Determine what period you should take data for. To do so, look at your answers to the previous questions. The data should allow you to build a model for 2017.Work only with the data that you've decided is relevant. Disregard the data for previous years.

Which platforms are leading in sales? Which ones are growing or shrinking? Select several potentially profitable platforms.


Build a box plot for the global sales of all games, broken down by platform. Are the differences in sales significant? What about average sales on various platforms? Describe your findings.

Take a look at how user and professional reviews affect sales for one popular platform (you choose). Build a scatter plot and calculate the correlation between reviews and sales. Draw conclusions.

Keeping your conclusions in mind, compare the sales of the same games on other platforms.

Take a look at the general distribution of games by genre. What can we say about the most profitable genres? Can you generalize about genres with high and low sales?

<h2>Create User Profile for each region</h2>

NA: 
    1. top 5 platforms and describe variantions in their market shares from regian to region.
    2. top 5 generes  and explain the difference.
    3. Do ESRB ratings affect sales in each region?

EU: 
    1. top 5 platforms and describe variantions in their market shares from regian to region.
    2. top 5 generes  and explain the difference.
    3. Do ESRB ratings affect sales in each region?

JP: 
    1. top 5 platforms and describe variantions in their market shares from regian to region.
    2. top 5 generes  and explain the difference.
    3. Do ESRB ratings affect sales in each region?

<h2>Test The Hypotheses</h2>
1- average user ratings of the xbox one and PC platforms are the same
2- average user ratings of the action and sports genres are the same
<i>set the alpha threshold value yourself</i>
Explain:
- how you formulated your hypotheses: null and alternative hypothesis
- what significance level you chose to test the hypotheses, and why

<h2>Write a Conclusion</h2>
- Data description
  - Name
  - Platform
  - Year_of_Release
  - Genre
  - NA_Sales(USD millions)
  - EU_Sales(USD millions)
  - JP_Sales(USD millions)
  - Other_Sales(USD millions)
  - Critic_Score(max 100)
  - User_Score(max 10)
  - Rating(ESRB)